File: generated_code/cpp/solution_141_1.cpp
#include <iostream>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        // Floyd's Tortoise and Hare Algorithm (Two Pointers)
        // Time Complexity: O(N) where N is the number of nodes.
        // Space Complexity: O(1) auxiliary space.
        
        // Edge case: An empty list or a list with only one node cannot have a cycle.
        if (!head || !head->next) {
            return false;
        }
        
        ListNode *slow = head;
        ListNode *fast = head;
        
        // Move slow pointer by 1 step and fast pointer by 2 steps.
        // If there is a cycle, they will eventually meet.
        // If there is no cycle, fast pointer will hit the end of the list (NULL).
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            
            // Cycle detected
            if (slow == fast) {
                return true;
            }
        }
        
        // Fast reached the end, so no cycle exists.
        return false;
    }
};

File: generated_code/python/solution_141_1.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Using Floyd's Tortoise and Hare algorithm (Two Pointers)
        # Time Complexity: O(N) where N is the number of nodes
        # Space Complexity: O(1) auxiliary space
        
        if not head or not head.next:
            return False
            
        slow = head
        fast = head
        
        while fast and fast.next:
            slow = slow.next          # Moves 1 step at a time
            fast = fast.next.next     # Moves 2 steps at a time
            
            # If there is a cycle, the fast pointer will eventually catch up to the slow pointer
            if slow == fast:
                return True
                
        # If fast reaches the end of the list, no cycle exists
        return False